Expand pyproject.toml guide to also cover the [build-system] table (#1392)
pypa/packaging.python.org: source/guides/writing-pyproject-toml.rst

Writing your ``pyproject.toml``
===============================

``pyproject.toml`` is a configuration file used by packaging tools, as
well as other tools such as linters, type checkers, etc. There are
three possible TOML tables in this file.

- The ``[build-system]`` table is **strongly recommended**. It allows
  you to declare which :term:`build backend` you use and which other
  dependencies are needed to build your project.

- The ``[project]`` table is the format that most build backends use to specify
  your project's basic metadata, such as the dependencies, your name, etc.

- The ``[tool]`` table has tool-specific subtables, e.g., ``[tool.hatch]``,
  ``[tool.black]``, ``[tool.mypy]``. We only touch upon this table here because
  its contents are defined by each tool. Consult the particular tool's
  documentation to know what it can contain.

.. note::

   There is a significant difference between the ``[build-system]`` and
   ``[project]`` tables. The former should always be present, regardless of
   which build backend you use (since it *defines* the tool you use). The latter
   is understood by *most* build backends, but some build backends use a
   different format.

>    At the time of this writing (November 2023), Poetry_ is a notable build

Reviewer: ```suggestion
   At the time of writing this (November 2023), Poetry_ is a notable build
```

Writing your ``pyproject.toml``
===============================

``pyproject.toml`` is a configuration file used by packaging tools, as
well as other tools such as linters, type checkers, etc. There are
three possible TOML tables in this file.

- The ``[build-system]`` table is **strongly recommended**. It allows
  you to declare which :term:`build backend` you use and which other
  dependencies are needed to build your project.

- The ``[project]`` table is the format that most build backends use to specify
  your project's basic metadata, such as the dependencies, your name, etc.

> - The ``[tool]`` table has arbitrary subtables corresponding to tools, whether

Reviewer 2: ```suggestion
- The ``[tool]`` table has tool specific subtables, whether
```

  packaging-related or not, e.g., ``[tool.hatch]``, ``[tool.black]``,
  ``[tool.mypy]``. We only touch upon this table here because its contents are
  entirely tool-specific. Consult each tool's documentation to know what it can
  contain.

.. note::

   There is a significant difference between the ``[build-system]`` and
   ``[project]`` tables. The former should always be present, regardless of
   which build backend you use (since it *defines* the tool you use). The latter
   is understood by *most* build backends, but some build backends use a
   different format.

   At the time of this writing (November 2023), Poetry_ is a notable build
   backend that does not use the ``[project]`` table (it uses the
   ``[tool.poetry]`` table instead).

   Also, the setuptools_ build backend supports both the ``[project]`` table,
   and the older format in ``setup.cfg`` or ``setup.py``. For new projects, it
   is recommended to use the ``[project]`` table, and keep ``setup.py`` only if

>    some programmatic configuration is needed (especially building C extensions),

Reviewer 2: ```suggestion
   some programmatic configuration is needed (such as building C extensions),
```

The ``[build-system]`` table contains a ``build-backend`` key, which specifies
the build backend to be used. It also contains a ``requires`` key, which is a
list of dependencies needed to build the project -- this is typically just the
build backend package, but it may also contain additional dependencies. You can
also constrain the versions, e.g., ``requires = ["setuptools >= 61.0"]``.

Usually, you'll just copy what your build backend's documentation
suggests (after :ref:`choosing your build backend <choosing-build-backend>`).
Here are the values for some common build backends:

.. tab:: Hatchling

    .. code-block:: toml

        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"

.. tab:: setuptools

    .. code-block:: toml

        [build-system]
        requires = ["setuptools >= 61.0"]
        build-backend = "setuptools.build_meta"

.. tab:: Flit

    .. code-block:: toml

        [build-system]
        requires = ["flit_core >= 3.4"]
        build-backend = "flit_core.buildapi"

.. tab:: PDM

    .. code-block:: toml

        [build-system]
        requires = ["pdm-backend"]
        build-backend = "pdm.backend"

> The rest of this guide is devoted to the ``[project]`` table.

Reviewer 2: Perhaps move this sentence to the start of the next section. It will be more readable for a new user starting the next section with the project table as is done for the  build-system table.
